### PR TITLE
PLT-1030 Implemented SOPs for UPLC

### PR DESCRIPTION
**Pull request by mjaskelioff**
Implemented Sum of products for `plc-agda`'s UPLC.

All tests in the conformance test pass.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Changelog fragments have been written (if appropriate)
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting master unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-metatheory/src/Untyped/CEK.lagda.md [hunk 1]
```diff
@@ -47,6 +47,7 @@ data Value where
   V-ƛ : ∀{X} → Env X → Maybe X ⊢ → Value
   V-con : (ty : TyTag) → ⟦ ty ⟧tag → Value
   V-delay : ∀{X} → Env X → X ⊢ → Value
+  V-constr : (i : ℕ) → (vs : List Value) → Value
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-metatheory/src/Untyped/CEK.lagda.md:50]: ... which it isn't in the real CEK machine. Unclear if that matters. And we might end up switching it later depending.

File: plutus-metatheory/src/Untyped/CEK.lagda.md [hunk 2]
```diff
@@ -405,30 +417,48 @@ BUILTIN' b {pt = pt} {pa = pa} bt with trans (sym (+-identityʳ _)) (∔2+ pt) |
 ival : Builtin → Value
 ival b = V-I b base
 
+extendStack : Stack → List Value → Stack
```
Comment (michaelpj) [inline plutus-metatheory/src/Untyped/CEK.lagda.md:420]: overly generic name?

File: plutus-metatheory/src/Untyped/CEK.lagda.md [hunk 3]
```diff
@@ -405,30 +417,48 @@ BUILTIN' b {pt = pt} {pa = pa} bt with trans (sym (+-identityʳ _)) (∔2+ pt) |
...
+extendStack s [] = s
+extendStack s (v ∷ vs) = extendStack (s , -·v v) vs
+
+lookup? : ∀{A} → ℕ → List A → Maybe A
```
Comment (michaelpj) [inline plutus-metatheory/src/Untyped/CEK.lagda.md:424]: not in the stdlib?

File: plutus-metatheory/src/Untyped/CEK.lagda.md [hunk 4]
```diff
@@ -405,30 +417,48 @@ BUILTIN' b {pt = pt} {pa = pa} bt with trans (sym (+-identityʳ _)) (∔2+ pt) |
...
+step ((s , force-) ◅ V-IΠ b bapp)          = s ◅ V-I b (app⋆ bapp)
+step ((s , force-) ◅ V-delay ρ t)          = step (s ; ρ ▻ t)
+step ((s , force-) ◅ V-ƛ _ _)              = ◆ -- lambda in delay position
+step ((s , force-) ◅ V-con _ _)            = ◆ -- constant in delay position
+step ((s , force-) ◅ V-I⇒ b bapp)          = ◆ -- function in delay position
+step ((s , force-) ◅ V-constr i vs)        = ◆ -- SOP in delay position
+step ((s , constr- i vs ρ []) ◅ v)         = s ◅ V-constr i (v ∷ vs)
+step ((s , constr- i vs ρ (x ∷ ts)) ◅ v)   = (s , constr- i (v ∷ vs) ρ ts); ρ ▻ x
+step ((s , case- ρ ts) ◅ V-constr i vs)    = maybe (extendStack s vs ; ρ ▻_) ◆ (lookup? i ts)
+step ((s , case- _ _ ) ◅ v)                = ◆ -- case of not a SOP
```
Comment (michaelpj) [inline plutus-metatheory/src/Untyped/CEK.lagda.md:459]: Haven't read this function before, so slightly off-topic, but I'm not sure what your approach is here: sometimes you list out all the "wrong" cases explicitly (like for all the cases for function application and forcing), sometimes you use a fall-through pattern like here. Is there a reason for the difference?

File: plutus-metatheory/src/Untyped/RenamingSubstitution.lagda.md [hunk 5]
```diff
@@ -28,15 +29,21 @@ lift : {X Y : Set} → Ren X Y → Ren (Maybe X) (Maybe Y)
...
+ren ρ (force t)     = force (ren ρ t)
+ren ρ (delay t)     = delay (ren ρ t)
+ren ρ (con tcn)     = con tcn
+ren ρ (builtin b)   = builtin b
+ren ρ error         = error
+ren ρ (constr i xs) = constr i (renList ρ xs)
+ren ρ (case x ts)   = case (ren ρ x) (renList ρ ts)
+
+renList ρ [] = []
+renList ρ (x ∷ xs) = ren ρ x ∷ renList ρ xs
```
Comment (michaelpj) [inline plutus-metatheory/src/Untyped/RenamingSubstitution.lagda.md:46]: Can one not just do `fmap (ren rho)`?

File: plutus-metatheory/src/Untyped/RenamingSubstitution.lagda.md [hunk 6]
```diff
@@ -54,19 +61,30 @@ lift-cong : ∀{X Y}
 lift-cong p nothing  = refl
 lift-cong p (just x) = cong just (p x)
 
+renList-cong : ∀{X Y}
+    {ρ ρ' : Ren X Y}
+  → (∀ x → ρ x ≡ ρ' x)
+  → (t : List (X ⊢))
+  → renList ρ t ≡ renList ρ' t
```
Comment (michaelpj) [inline plutus-metatheory/src/Untyped/RenamingSubstitution.lagda.md:68]: Ah I see, we need the function so we can prove this lemma about it?

File: plutus-metatheory/src/Untyped/RenamingSubstitution.lagda.md [hunk 7]
```diff
@@ -54,19 +61,30 @@ lift-cong : ∀{X Y}
...
+ren-cong p (force t)     = cong force (ren-cong p t)
+ren-cong p (delay t)     = cong delay (ren-cong p t)
+ren-cong p (con c)       = refl
+ren-cong p (builtin b)   = refl
+ren-cong p error         = refl
+ren-cong p (constr i xs) = cong (constr i) (renList-cong p xs)
+ren-cong p (case t ts)   = cong₂ case (ren-cong p t) (renList-cong p ts)
+
+renList-cong p [] = refl
+renList-cong p (x ∷ xs) = cong₂ _∷_ (ren-cong p x) (renList-cong p xs)
```
Comment (michaelpj) [inline plutus-metatheory/src/Untyped/RenamingSubstitution.lagda.md:87]: Somehow seems like you should be able to get this automatically by the proof-equivalent of `fmap`, but whatever.

**Review by michaelpj — APPROVED** [review]

File: plutus-metatheory/src/Untyped.hs [hunk 8]
```diff
@@ -22,7 +22,7 @@ data UTerm = UVar Integer
            | UBuiltin DefaultFun
            | UDelay UTerm
            | UForce UTerm
-           | UConstr Word64 [UTerm]
+           | UConstr Integer [UTerm]
```
[the remark this answers is not in the document]
Comment (mjaskelioff) [reply plutus-metatheory/src/Untyped.hs:25]: Agda compiles $\mathbb{N}$ to a Haskell `Integer`, so I need this type to be `Integer`.

File: plutus-metatheory/src/Untyped/CEK.lagda.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjaskelioff) [reply plutus-metatheory/src/Untyped/CEK.lagda.md:50]: I could put it the other way but it's more efficient like this. Another option would be to use some snoc list instead.

File: plutus-metatheory/src/Untyped/RenamingSubstitution.lagda.md (hunk 5 above)
Comment (mjaskelioff) [reply plutus-metatheory/src/Untyped/RenamingSubstitution.lagda.md:46]: I wish I could, my job would be so much easier. However, the termination checker doesn't like it. AFAIK, the only way around the termination checker is to write a mutually recursive specialised map function, like it's done here.

File: plutus-metatheory/src/Untyped/CEK.lagda.md (hunk 2 above)
Comment (mjaskelioff) [reply plutus-metatheory/src/Untyped/CEK.lagda.md:420]: Agree. Will change it to `pushValueFrames`.

File: plutus-metatheory/src/Untyped/CEK.lagda.md (hunk 4 above)
Comment (mjaskelioff) [reply plutus-metatheory/src/Untyped/CEK.lagda.md:459]: I don't really have an answer. I guess the previous definitions were listed explicitly to make sure that all the good cases are being handled? I should probably follow the existing style and list all cases...

File: plutus-metatheory/src/Untyped/RenamingSubstitution.lagda.md (hunk 6 above)
Comment (mjaskelioff) [reply plutus-metatheory/src/Untyped/RenamingSubstitution.lagda.md:68]: Same case of needing to satisfy the Termination Checker.

**Pull request merged**